### ISSUE TITLE
Timeout

### ISSUE THREAD

**Issue by christianostrem** [oidc-wp/openid-connect-generic#310]
Hi, 

I am working on an integration with idporten. A goverment server for openid.

The problem is that i get a timeout in this function:

File:
openid-connect-generic-client.php

Function:
request_authentication_token

method: $response = wp_remote_post( $this->endpoint_token, $request );

Steps: 
1) I click the login button. -> Next a window appears . Then i login. All ok.
2) The the oauth server redirects to wordpress. It reaches the request_authentication_token function and runs the wp_remote_post( $this->endpoint_token, $request ) but it times out.

This function is triggerd: if ( is_wp_error( $response ) )

**Expected behavior**
I expected to receive a token

**Isolating the problem (mark completed items with an [x]):**
- [ X] I have deactivated other plugins and confirmed this bug occurs when only this plugin is active.
- [ X] This bug happens with a default WordPress theme active.
- [ X] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
- Website URL: 
- PHP Version: 7.4.14
- WordPress Version: 5.7.1
- Plugin Version: 3.8.5
- Identity Provider: Idporten
- Relevant Plugin Settings: 

  Alternate Redirect URI since the openid server dosent support the querystring parameters.
  client_secret_post
  authorization_code 
  --
  Additional header parameters
  'client_orgno' => 00000000,
  'acr_values' => 'Level3',
  'ui_locales' => 'nb',
  --
  Added PKCE support.
  code_verifier, code_challenge_method and code_challenge

Anything that could point me in the right direction would be great.

Docs:
https://docs.digdir.no/oidc_guide_english.html
Ps: yes it supports: client_secret_post

**Comment by timnolte** [comment]
@christianostrem so it sounds like you've modified the plugin code, is that correct? If there is a timeout occuring then that means there is an issue with the request your are sending to the IDP. I need to know what code you've changed. However, since you've made your own code changes in order to get the plugin to run with your specific IDP then there is probably not much support we can give.

**Comment by christianostrem** [comment]
Yes, i did a modification, but i still get the same timeout without the modification.

These are the modifications i did:

`
in file: openid-connect-generic-client.php
in function: request_authentication_token

$request = array(
	'body' => array(
		'code'          => $code,
		'client_id'     => $this->client_id,
		'client_secret' => $this->client_secret,
		'redirect_uri'  => $this->redirect_uri,
		'grant_type'    => 'authorization_code',
		'scope'         => $this->scope,
              //Added these parameters
              'client_orgno' => 00000000,
              'acr_values' => 'Level3',
              'ui_locales' => 'nb',
              'code_verifier' => 'xxxxx' //PKCE code_verifier
	),
	'headers' => array( 'Host' => $host ),
);`

`
in file: openid-connect-generic-client-wrapper.php
in function: get_authentication_url

Added the code_challenge_method and code_challenge

$url = sprintf(
'%1$s%2$sresponse_type=code&scope=%3$s&client_id=%4$s&state=%5$s**&code_challenge_method=S256&
code_challenge=xxxxx**&redirect_uri=%6$s',
	$atts['endpoint_login'],
	$separator,
	rawurlencode( $atts['scope'] ),
	rawurlencode( $atts['client_id'] ),
	$this->client->new_state( $atts['redirect_to'] ),
	rawurlencode( $atts['redirect_uri'] )
);

`

**Comment by christianostrem** [comment]
I have had new development her. Ill come back to this ticket within a week if its still relevant.

**Comment by christianostrem** [comment]
Still relevant.

**Comment by christianostrem** [comment]
Iam now running a non modified version of the plugin, and its still doing timeout.

Relevant documentation authorize:
https://docs.digdir.no/oidc_protocol_authorize.html

Error in log:

Type: make_authentication_urlDate: 2021-05-03 12:12:30User: 0URI : /wp-login.php?login-error=http_request_failed&message=cURL+error+28%3A+Operation+timed+out+after+15001+milliseconds+with+0+bytes+received | string(268) "https://oidc-ver2.difi.no/idporten-oidc-provider/authorize?response_type=code&scope=openid%20profile&client_id=ec1a552a-c792-4609-ae33-f8db30246291&state=2a0751d7f52fdcab6134325b0b2bb101&redirect_uri=https%3A%2F%2Fxxx.com%2Fopenid-connect-authorize"


http_request_failedDate: 2021-05-03 12:17:42User: 0URI : /openid-connect-authorize?code=&state=d5566bd1504958c25be0ad3d4d29ae0d | object(WP_Error)#18217 (3) { ["errors"]=> array(2) { ["http_request_failed"]=> array(1) { [0]=> string(81) "cURL error 28: Operation timed out after 15001 milliseconds with 0 bytes received" } ["request_authentication_token"]=> array(1) { [0]=> string(40) "Request for authentication token failed." } } ["error_data"]=> array(0) { } ["additional_data":protected]=> array(0) { } }

this is the last ting logged before timeout

**Comment by christianostrem** [comment]
I have finally found out what the problem is. The plugin is using the wp_remote_post function and its someting with that function that dosent work. I replaced it with file_get_content instead and then it worked perfectly. One problem  with file_get_content is that it dosent have very good error handling. I have tried different options in wp_remote_post also but without success. Any suggestions to make wp_remote_post work is welcome.

**Comment by daggerhart** [comment]
Hi @christianostrem,
I've no idea if this is relevant, but yesterday I ran into an issue w/ `wp_remote_X` functions with another IDP. The IDP did not like that user-agent that the wp functions were adding to every call. 

On the offchance this is relevant to your issue, could you try adding `'user-agent' => false,` to the `$request` array in these two functions? Not as a header or in the body, but within the top level of the $request array:
* https://github.com/oidc-wp/openid-connect-generic/blob/dev/includes/openid-connect-generic-client.php#L197
* https://github.com/oidc-wp/openid-connect-generic/blob/dev/includes/openid-connect-generic-client.php#L236

**Comment by timnolte** [comment]
Yeah, so the `wp_remote_post` method uses cURL under the hood. It still seems like there is an issue with your IDP responding. You can use the `http_request_timeout` filter hook to change the timeout to something longer to account for slow processing/response by your IDP.

**Comment by christianostrem** [comment]
isn't http_request_timeout the same as the timeout set in the plugin?

**Comment by christianostrem** [comment]
I have found the reason why it dosent work.
After a lot of testing it seems like they are blocking the user-agent: wordpress ...

By adding a hook i have changed the user-agent and now everything works.
